pkg/templates/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -0,0 +1,201 @@
+	revisions map[string]map[string]string
+	decoder   runtime.Decoder
+}
+
+// NewNSTemplateTierGenerator returns a new NSTemplateTierGenerator
+func NewNSTemplateTierGenerator(s *runtime.Scheme, asset func(name string) ([]byte, error)) (*NSTemplateTierGenerator, error) {
+	metadata, err := asset("metadata.yaml")
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize the NSTemplateTierGenerator")
+	}
+	revisions, err := parseAllRevisions(metadata)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize the NSTemplateTierGenerator")
+	}
+	return &NSTemplateTierGenerator{
+		asset:     asset,
+		revisions: revisions,
+		decoder:   serializer.NewCodecFactory(s).UniversalDeserializer(),
+	}, nil
+}
+
+// parseRevisions returns a "supermap" in which:
+// - each key is a tier kind (eg: "basic", "advanced", etc.)
+// - each value is a map of revisions indexed by their associated template kind
+//   (eg: {"code":"123456", "dev":"abcdef", "stage":"cafe01"})
+func parseAllRevisions(metadata []byte) (map[string]map[string]string, error) {
+	data := make(map[string]interface{})
+	err := yaml.Unmarshal([]byte(metadata), &data)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse all template revisions")
+	}
+	// reads all entries in the YAML, splitting the keys
+	// as we expect the following format: '<tier_kind>-<namespace_kind>' (eg: 'advanced-code')
+	revisions := make(map[string]map[string]string)
+	for filename, revision := range data {
+		data := strings.Split(filename, "-")
+		if len(data) != 2 {
+			log.Info("invalid namespace template filename. Expected format: '<tier_kind>-<namespace_kind>'", "filename", filename)
+			continue
+		}
+		tierKind := data[0]
+		nsKind := data[1]
+		// create a new entry if needed
+		revision, ok := revision.(string)
+		if !ok {
+			log.Info("invalid namespace template filename revision. Expected a string", "filename", filename, "revision", revision)
+			continue
+		}
+		if _, ok := revisions[tierKind]; !ok {
+			revisions[tierKind] = make(map[string]string, 3) // expect 3 entries: 'code', 'dev' and 'stage'
+		}
+		revisions[tierKind][nsKind] = revision
+	}
+	log.Info("templates revisions loaded", "revisions", revisions)
+	return revisions, nil
+}
+
+// NewNSTemplateTiers generates all manifests, indexed by their associated tier kind and by their namespace kind
+// eg:
+// - advanced:
+//   - code: <[]byte>
+//   - dev: <[]byte>
+//   - stage: <[]byte>
+// - basic:
+//   - code: <[]byte>
+//   - dev: <[]byte>
+//   - stage: <[]byte>
+func (g NSTemplateTierGenerator) NewNSTemplateTiers(namespace string) ([]*toolchainv1alpha1.NSTemplateTier, error) {
+	objects := make([]*toolchainv1alpha1.NSTemplateTier, 0, len(g.revisions))
+	// retrieve the tier names and order them, so we compare them
+	// with the expected templates during the tests
+	tiers := make([]string, 0, len(g.revisions))
+	for tier := range g.revisions {
+		tiers = append(tiers, tier)
+	}
+	sort.Strings(tiers)
```

Comment:
I guess that he update is not implemented yet (which is fine) but I really don't see the purpose of this sorting. We should go through the tiers one by one and compare the particular newly generated  `NSTemplateTier` with the one existing in the cluster and update the particular one if needed. I think that there is no need for keeping the order.

Reply:
the sorting here helped with the comparison in the test, but I guess I could return a map instead of a slice, so there would be no need for such sorting. Let me do that

Reply:
done in d68ff4e